### DISCUSSION TITLE
Not able to run the State Machine and FinitStateMachine Example

### DISCUSSION THREAD

**Discussion by Gregor841** [FreeOpcUa/opcua-asyncio#1572] [Q&A]
I am quite new to this library and was trying some of the examples successfully but when I tried to run the StateMachine (statemachine-example.py) example, I got the following Error in line 52: 

49        # initialise the StateMachine by call change_state() with the InitialState
50        # if the statechange should trigger an TransitionEvent the Message can be assigned here
51        # if event_msg is None no event will be triggered
52         await mystatemachine.change_state(state1, trans1, f"{mystatemachine._name}: Idle", 300)

asyncua.ua.uaerrors._auto.BadTypeMismatch: The value supplied for the attribute is not of the same type as the attribute"s value.(BadTypeMismatch)

By commenting out all the change_state calls, the code runs till the end but then obviously the  state machine is not doing anything anymore. 

I also tried to use the finiteStateMachine (finitestatemachine-example.py) 
but there I get the following Error in line 19

17        mystatemachine = FiniteStateMachine(server, server.nodes.objects, idx, "FiniteStateMachine")
18        # call statemachine.install() to instantiate the statemachinetype (with or without optional nodes)
19        await mystatemachine.install(optionals=True)

InstantiationError NodeId: NodeId(Identifier=2771, NamespaceIndex=0, NodeIdType=<NodeIdType.FourByte: 1>) is abstract and cant be instantiated!

By checking the OPC UA FiniteStateMachineType (Node id 2771) it is specified that indeed this type **IsAbstract**
Therefore I am also stuck with this example. 

I am using Python 3.12.1 for both examples

Could you please help me further and point me in a direction on how to solve the issues and run the StateMachine example? Many thanks in advance.

**Comment by AndreasHeine** [comment]
i will take a look tomorrow!

**Reply by AndreasHeine** [reply]
statemachine-example:

looks like a bug...

```
WARNING:asyncua.server.address_space:Write refused: Variant: Variant(Value=NodeId(Identifier=6, NamespaceIndex=2, NodeIdType=<NodeIdType.FourByte: 1>), VariantType=<VariantType.NodeId: 17>, Dimensions=None, is_array=False) with type VariantType.NodeId does not have expected type: VariantType.Null
Traceback (most recent call last):
  File "/home/andreas/GitHub/opcua-asyncio/examples/statemachine-example.py", line 85, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/andreas/GitHub/opcua-asyncio/examples/statemachine-example.py", line 52, in main
    await mystatemachine.change_state(state1, trans1, f"{mystatemachine._name}: Idle", 300)
  File "/home/andreas/.local/lib/python3.10/site-packages/asyncua/common/statemachine.py", line 183, in change_state
    await self._write_state(state)
  File "/home/andreas/.local/lib/python3.10/site-packages/asyncua/common/statemachine.py", line 206, in _write_state
    await self._current_state_id_node.write_value(state.node.nodeid, varianttype=ua.VariantType.NodeId)
  File "/home/andreas/.local/lib/python3.10/site-packages/asyncua/common/node.py", line 269, in write_value
    await self.write_attribute(ua.AttributeIds.Value, dv)
  File "/home/andreas/.local/lib/python3.10/site-packages/asyncua/common/node.py", line 323, in write_attribute
    result[0].check()
  File "/home/andreas/.local/lib/python3.10/site-packages/asyncua/ua/uatypes.py", line 375, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadTypeMismatch: The value supplied for the attribute is not of the same type as the attribute"s value.(BadTypeMismatch)
```

type VariantType.NodeId does not have expected type: VariantType.Null

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/8cf8682a019bacc7e67c42fec72531bf3e87c31e/asyncua/server/address_space.py#L814

```python
        if vtype == ua.VariantType.Null:
            # Node had a null value, many nodes are initialized with that value
            # we should check what the real type is
            dtype = node.attributes[ua.AttributeIds.DataType].value.Value.Value  # type: ignore[union-attr]
            if dtype.NamespaceIndex == 0 and dtype.Identifier <= 25:
                vtype = ua.VariantType(dtype.Identifier)
            else:
                # FIXME: should find the correct variant type given data type but
                # this is a bit complicaed so trusting the first write
                return True
        if value.Value.VariantType == vtype:  # type: ignore[union-attr]
            return True
        _logger.warning(
            "Write refused: Variant: %s with type %s does not have expected type: %s",
            value.Value,
            value.Value.VariantType if value.Value else None,
            attval.value.Value.VariantType if attval.value.Value else None,
        )
        return False
```

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/Core/Part16/v105/docs/4.4.3

HasProperty | Variable | Id | BaseDataType | PropertyType | Mandatory

according to spec it has BaseDataType which means it can be any Subtype of it!

> [Id ](https://reference.opcfoundation.org/search/309?t=Id)is a name which uniquely identifies the current state within the [StateMachineType](https://reference.opcfoundation.org/search/309?t=StateMachineType). A subtype may restrict the [DataType](https://reference.opcfoundation.org/search/309?t=DataType).

which just telling its a name, but it does not tell that its a specific DataType...

> A subtype may restrict the [DataType](https://reference.opcfoundation.org/search/309?t=DataType).

StateVariableType is not an abstract type so it can be used as it is... with BaseDataType as DataType


asyncua initialize it as DataType String
node-opcua initialize it as DataType "empty" (ns=0;i=0)


@oroulet can you remember why this got changed and why we initialize it as string?
